platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StartupStateUtils.java

```diff
@@ -0,0 +1,586 @@
+     *
+     * @param platformContext        the platform context
+     * @param recycleBin             the recycle bin
+     * @param currentSoftwareVersion the current software version
+     * @param savedStateFiles        the saved states to try
+     * @return the loaded state
+     */
+    private static ReservedSignedState loadLatestState(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final RecycleBin recycleBin,
+            @NonNull final SoftwareVersion currentSoftwareVersion,
+            @NonNull final List<SavedStateInfo> savedStateFiles)
+            throws SignedStateLoadingException {
+
+        logger.info(STARTUP.getMarker(), "Loading latest state from disk.");
+
+        for (final SavedStateInfo savedStateFile : savedStateFiles) {
+            final ReservedSignedState state =
+                    loadStateFile(platformContext, recycleBin, currentSoftwareVersion, savedStateFile);
+            if (state != null) {
+                return state;
+            }
+        }
+
+        logger.warn(STARTUP.getMarker(), "No valid saved states were found on disk. Starting from genesis.");
+        return createNullReservation();
+    }
+
+    /**
+     * Load the requested state from file. If state can not be loaded, recycle the file and return null.
+     *
+     * @param platformContext        the platform context
+     * @param recycleBin             the recycle bin
+     * @param currentSoftwareVersion the current software version
+     * @param savedStateFile         the state to load
+     * @return the loaded state, or null if the state could not be loaded. Will be fully hashed if non-null.
+     */
+    @Nullable
+    private static ReservedSignedState loadStateFile(
+            @NonNull final PlatformContext platformContext,
+            @NonNull final RecycleBin recycleBin,
+            @NonNull final SoftwareVersion currentSoftwareVersion,
+            @NonNull final SavedStateInfo savedStateFile)
+            throws SignedStateLoadingException {
+
+        logger.info(STARTUP.getMarker(), "Loading signed state from disk: {}", savedStateFile.stateFile());
+
+        final DeserializedSignedState deserializedSignedState;
+        try {
+            deserializedSignedState = readStateFile(platformContext, savedStateFile.stateFile());
+        } catch (final IOException e) {
+            logger.error(EXCEPTION.getMarker(), "unable to load state file {}", savedStateFile.stateFile(), e);
+
+            final StateConfig stateConfig = platformContext.getConfiguration().getConfigData(StateConfig.class);
+            if (stateConfig.deleteInvalidStateFiles()) {
+                recycleState(recycleBin, savedStateFile);
+                return null;
+            } else {
+                throw new SignedStateLoadingException("unable to load state, this is unrecoverable");
+            }
+        }
+
+        final State state = deserializedSignedState.reservedSignedState().get().getState();
```

Comment:
Does this reserved signed state not need closing?

Reply:
It doesn't need closing because we pass the reservation to the caller of the method.